components/raftstore/src/store/async_io/write_router.rs

```diff
@@ -223,54 +227,91 @@ where
+                if write_senders.send(self.writer_id, msg).is_err() {
                     // Write threads are destroyed after store threads during shutdown.
-                    panic!("{} failed to send write msg, err: disconnected", self.tag);
+                    safe_panic!("{} failed to send write msg, err: disconnected", self.tag);
                 }
                 ctx.raft_metrics()
                     .write_block_wait
                     .observe(now.saturating_elapsed_secs());
             }
             Err(TrySendError::Disconnected(_)) => {
                 // Write threads are destroyed after store threads during shutdown.
-                panic!("{} failed to send write msg, err: disconnected", self.tag);
+                safe_panic!("{} failed to send write msg, err: disconnected", self.tag);
             }
         }
     }
 }
 
+pub type SenderVec<EK, ER> = Vec<Sender<WriteMsg<EK, ER>>>;
+
 /// Senders for asynchronous writes. There can be multiple senders, generally
 /// you should use `WriteRouter` to decide which sender to be used.
 #[derive(Clone)]
 pub struct WriteSenders<EK: KvEngine, ER: RaftEngine> {
-    write_senders: Vec<Sender<WriteMsg<EK, ER>>>,
+    senders: Tracker<SenderVec<EK, ER>>,
+    cached_senders: SenderVec<EK, ER>,
     io_reschedule_concurrent_count: Arc<AtomicUsize>,
 }
 
 impl<EK: KvEngine, ER: RaftEngine> WriteSenders<EK, ER> {
-    pub fn new(write_senders: Vec<Sender<WriteMsg<EK, ER>>>) -> Self {
+    pub fn new(senders: Arc<VersionTrack<SenderVec<EK, ER>>>) -> Self {
+        let cached_senders = senders.value().clone();
         WriteSenders {
-            write_senders,
+            senders: senders.tracker("async writers' trackers".to_owned()),
+            cached_senders,
             io_reschedule_concurrent_count: Arc::default(),
         }
     }
 
     #[inline]
     pub fn is_empty(&self) -> bool {
-        self.write_senders.is_empty()
+        self.cached_senders.is_empty()
+    }
+
+    #[inline]
+    /// Returns the valid capacity of async senders.
+    pub fn capacity(&self) -> usize {
+        self.cached_senders.len()
     }
 }
 
-impl<EK: KvEngine, ER: RaftEngine> Index<usize> for WriteSenders<EK, ER> {
-    type Output = Sender<WriteMsg<EK, ER>>;
+impl<EK: KvEngine, ER: RaftEngine> WriteSenders<EK, ER> {
+    #[inline]
+    pub fn refresh(&mut self) {
+        if let Some(senders) = self.senders.any_new() {
+            self.cached_senders = senders.clone();
+        }
+    }
+
+    #[inline]
+    pub fn send(
+        &self,
+        idx: usize,
+        msg: WriteMsg<EK, ER>,
+    ) -> Result<(), SendError<WriteMsg<EK, ER>>> {
+        debug_assert!(!self.is_empty());
+        // Because the given `idx` is not atomically updated during the progress of
+        // resizing, there exists one corner case should be tackled. That is, the given
+        // `idx` for accessing the senders might be obsolete when resizing the
```

Comment:
How? Isn't the raft pool stopped?

Reply:
It's inelegant here. And I've polished it by moving the checking when doing `should_send` check.

And about the corner case -- 
[prerequisite]
when we trying to pause the poller threads to stop the writing for a while, there still might have some events already received by pollers, and these events will be counted into `left_fsm_cnt` and dispatched to schedulers.

https://github.com/tikv/tikv/blob/65a99a89b9f03de1ca24cee8c33584d13370becc/components/batch-system/src/batch.rs#L504-L517

When we try to `decrease async-ios`, if `left_fsm_cnt` > 0, it means that still exists several `IOTask` should be tackled by schedulers, and should be sent to async-ios. In this case, the cached writers in schedulers haven't been refreshed, it might get an obsolete `idx`(generated by `randInt % (previous_capacity)`) for sending, causing `out of index` error. And I've update the annotation in `should_send` as followings shown:

https://github.com/LykxSassinator/tikv/blob/7f9fdeb06051b0ed586d3ada5c777b223c9c5ffe/components/raftstore/src/store/async_io/write_router.rs#L169-L184